Experimental/src/Ranges/Range.cs

```diff
@@ -0,0 +1,713 @@
+        }
+
+        public bool IsFull
+        {
+            get { return _options == RangeOptions.None; }
+        }
+
+        private RangeValue StartValue
+        {
+            get { return new RangeValue(Start, HasStart, IncludeStart); }
+        }
+
+        private RangeValue EndValue
+        {
+            get { return new RangeValue(End, HasEnd, IncludeEnd); }
+        }
+
+        public int CompareTo(Range<TValue> other)
+        {
+            var current = this;
+            if (!current.HasStart)
+				if (!other.HasStart)
+					return 0;
+				else
+					return -1;
+
+			if (!other.HasStart)
+                return 1;
+
+            var compare = current.Start.CompareTo(other.Start);
+            if (compare == 0)
+            {
+                if (current.HasEnd && other.HasEnd)
+                {
+                    compare = current.End.CompareTo(other.End);
+                    if (compare == 0)
+                    {
+                        if (current.IncludeEnd != other.IncludeEnd)
+                        {
+                            if (current.IncludeEnd)
+                                compare = 1;
+                            else
+                                compare = -1;
+                        }
+                    }
+                }
+            }
+
+            return compare;
+        }
+
+        public bool Equals(Range<TValue> other)
+        {
+            return _options == other._options
+                   && EqualityComparer<TValue>.Default.Equals(End, other.End) &&
+                   EqualityComparer<TValue>.Default.Equals(Start, other.Start);
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            return obj is Range<TValue> && Equals((Range<TValue>) obj);
+        }
+
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                var hashCode = (int) _options;
+                hashCode = (hashCode * 397) ^ EqualityComparer<TValue>.Default.GetHashCode(End);
+                hashCode = (hashCode * 397) ^ EqualityComparer<TValue>.Default.GetHashCode(Start);
+                return hashCode;
+            }
+        }
+
+        public Range<TValue> SetStart(TValue newStart, bool? include = null)
+        {
+            return new Range<TValue>(newStart, End,
+                ((include ?? IncludeStart) ? RangeOptions.IncludingStart : RangeOptions.None) |
+                (IncludeEnd ? RangeOptions.IncludingEnd : RangeOptions.None));
+        }
+
+        public Range<TValue> SetEnd(TValue newEnd, bool? include = null)
+        {
+            return new Range<TValue>(Start, newEnd,
+                (IncludeStart ? RangeOptions.IncludingStart : RangeOptions.None) |
+                ((include ?? IncludeEnd) ? RangeOptions.IncludingEnd : RangeOptions.None));
+        }
+
+        public Range<TValue> Union(Range<TValue> other)
```

Comment:
It's from WinApi, like UnionRect. It's smallest range that contains both ranges.